src/session/session-utils.c

```diff
@@ -70,23 +63,74 @@ read_authorize_response (const char *what)
        exit (0);
     }
 
+  /* make room for and add null terminator */
+  message = realloc (message, len + 1);
+  message[len] = '\0';
+  return (char *)message;
+}
+
+static const char *skip_quoted_value (const char *str)
+{
+  if (!*str || *str++ != '"')
+    errx (EX, "syntax error, expected quoted value");
+  while (*str && *str != '"')
+    str++;
+  if (*str != '"')
+    errx (EX, "syntax error, expected closing quote");
+  return str + 1;
+}
+
+char *get_authorize_key (const char *json, const char *key, bool required)
+{
   /*
-   * The authorize messages we receive always have an exact prefix and suffix:
+   * The authorize messages we receive always have this structure:
+   *
+   * \n{"command":"authorize","cookie":"NNN","response":"....","remote-peer":"..."}
    *
-   * \n{"command":"authorize","cookie":"NNN","response":"...."}
+   * We make many assumptions/restrictions here to keep the parser simple. In particular, no extra
+   * whitespace, everything is a string, no escaped ", and values don't contain commas.
    */
-  if (len <= auth_prefix_size + auth_response_size + auth_suffix_size ||
-      memcmp (message, auth_prefix, auth_prefix_size) != 0 ||
-      memcmp (message + auth_prefix_size, auth_response, auth_response_size) != 0 ||
-      memcmp (message + (len - auth_suffix_size), auth_suffix, auth_suffix_size) != 0)
+  if (!json || *json++ != '\n' || *json++ != '{')
+    errx (EX, "authorize message doesn't start with \\n{");
+
+  char *key_quoted;
+  int key_quoted_len = asprintf (&key_quoted, "\"%s\"", key);
+  if (key_quoted_len < 0)
+    errx (EX, "out of memory");
+
+  /* parse each "key":"value" pair until we find a matching key */
+  while (*json != '\0' && *json != '}')
```

Comment:
I like your approach of rescanning the entire string on every key lookup.  It keeps things a lot easier to reason about.  I would have probably tried to do something much more complicated.

If we're going to rescan the list for each key then why not just search for `,"${KEY}":"` using `strstr()` and consume the string from there, up to the next `'"'` (`strcspn()`)?  We can be 100% sure we're "on the right side" of the quoting because:
 - `,` can't appear inside of a message
 - a alphanumeric token can't possibly appear immediately adjacent to (the outside of) a double-quoted string: we'd expect to see some other kind of "punctuation" there.  In fact, given this fact, you could probably even drop the leading `,`.

Reply:
This is so much easier indeed. Rewritten. I did not match for the `,` as that's not true for the first entry in the object, and I don't think it's necessary.